### PR TITLE
hot-fix: paginate indexLookUp executor

### PR DESCRIPTION
**Pull request by wshwsh12**
<!-- Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?

Issue Number: close #xxx <!-- REMOVE this line if no issue to close -->

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) <!-- REMOVE this line if not applicable -->

What's Changed:

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note <!-- bugfixes or new feature need a release note -->

- No release note  <!-- Please write a release note here to describe the change you made when it is released to the users of TiDB. If your PR doesn't involve any change to TiDB(like test enhancements, RFC proposals...), you can write `No release note`. -->



### Benchmark

workload:
Create table t(a int, b int, c int, index idx(a,b));
random insert 1000000 rows
SQL: 
16 concurrency
```
[tidb]> desc analyze  select c from t where a < 9000 and c > 2000 order by a,b limit 100;
+--------------------------------------+---------+---------+-----------+--------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------+---------+------+
| id                                   | estRows | actRows | task      | access object            | execution info                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | operator info                      | memory  | disk |
+--------------------------------------+---------+---------+-----------+--------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------+---------+------+
| Projection_8                         | 100.00  | 100     | root      |                          | time:324.1ms, loops:2, Concurrency:OFF                                                                                                                                                                                                                                                                                                                                                                                                                                              | test.t.c                           | 3.38 KB | N/A  |
| └─Limit_13                           | 100.00  | 100     | root      |                          | time:324.1ms, loops:2                                                                                                                                                                                                                                                                                                                                                                                                                                                               | offset:0, count:100                | N/A     | N/A  |
|   └─Projection_35                    | 100.00  | 100     | root      |                          | time:324.1ms, loops:1, Concurrency:OFF                                                                                                                                                                                                                                                                                                                                                                                                                                              | test.t.a, test.t.b, test.t.c       | 4.50 KB | N/A  |
|     └─IndexLookUp_34                 | 100.00  | 100     | root      |                          | time:324.1ms, loops:1, index_task: {total_time: 325.7ms, fetch_handle: 321.8ms, build: 3.82ms, wait: 21.1µs}, table_task: {total_time: 1.7s, num: 7, concurrency: 5}                                                                                                                                                                                                                                                                                                                |                                    | 7.26 MB | N/A  |
|       ├─IndexRangeScan_31(Build)     | 122.32  | 933888  | cop[tikv] | table:t, index:idx(a, b) | time:320.3ms, loops:31, cop_task: {num: 1, max: 319.8ms, proc_keys: 933888, tot_proc: 310ms, rpc_num: 1, rpc_time: 319.7ms, copr_cache: disabled}, tikv_task:{time:304ms, loops:917}, scan_detail: {total_process_keys: 933888, total_keys: 933889, rocksdb: {delete_skipped_count: 0, key_skipped_count: 933888, block: {cache_hit_count: 581, read_count: 0, read_byte: 0 Bytes}}}                                                                                                | range:[-inf,9000), keep order:true | N/A     | N/A  |
|       └─Selection_33(Probe)          | 100.00  | 13280   | cop[tikv] |                          | time:78.8ms, loops:24, cop_task: {num: 12, max: 32.8ms, min: 2.71ms, avg: 9.59ms, p95: 32.8ms, max_proc_keys: 5103, p95_proc_keys: 5103, tot_proc: 94ms, rpc_num: 12, rpc_time: 114.8ms, copr_cache: disabled}, tikv_task:{proc max:30ms, min:0s, p80:13ms, p95:30ms, iters:66, tasks:12}, scan_detail: {total_process_keys: 15328, total_keys: 15328, rocksdb: {delete_skipped_count: 0, key_skipped_count: 0, block: {cache_hit_count: 4848, read_count: 0, read_byte: 0 Bytes}}} | gt(test.t.c, 2000)                 | N/A     | N/A  |
|         └─TableRowIDScan_32          | 122.32  | 15328   | cop[tikv] | table:t                  | tikv_task:{proc max:30ms, min:0s, p80:13ms, p95:30ms, iters:66, tasks:12}, scan_detail: {total_process_keys: 0, total_keys: 0, rocksdb: {delete_skipped_count: 0, key_skipped_count: 0, block: {cache_hit_count: 0, read_count: 0, read_byte: 0 Bytes}}}                                                                                                                                                                                                                            | keep order:false                   | N/A     | N/A  |
+--------------------------------------+---------+---------+-----------+--------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------+---------+------+
7 rows in set (0.373 sec)
```
After Optimization:
```
[tidb]> desc analyze  select c from t where a < 9000 and c > 2000 order by a,b limit 100;
+--------------------------------------+---------+---------+-----------+--------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------+---------+------+
| id                                   | estRows | actRows | task      | access object            | execution info                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | operator info                      | memory  | disk |
+--------------------------------------+---------+---------+-----------+--------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------+---------+------+
| Projection_8                         | 100.00  | 100     | root      |                          | time:18.6ms, loops:2, Concurrency:OFF                                                                                                                                                                                                                                                                                                                                                                                                                                                 | test.t.c                           | 3.38 KB | N/A  |
| └─Limit_13                           | 100.00  | 100     | root      |                          | time:18.6ms, loops:2                                                                                                                                                                                                                                                                                                                                                                                                                                                                  | offset:0, count:100                | N/A     | N/A  |
|   └─Projection_35                    | 100.00  | 100     | root      |                          | time:18.6ms, loops:1, Concurrency:OFF                                                                                                                                                                                                                                                                                                                                                                                                                                                 | test.t.a, test.t.b, test.t.c       | 4.50 KB | N/A  |
|     └─IndexLookUp_34                 | 100.00  | 100     | root      |                          | time:18.6ms, loops:1, index_task: {total_time: 20ms, fetch_handle: 16.4ms, build: 3.55ms, wait: 19.6µs}, table_task: {total_time: 204.8ms, num: 8, concurrency: 5}                                                                                                                                                                                                                                                                                                                    |                                    | 1.21 MB | N/A  |
|       ├─IndexRangeScan_31(Build)     | 122.32  | 20001   | cop[tikv] | table:t, index:idx(a, b) | time:15.6ms, loops:25                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | range:[-inf,9000), keep order:true | N/A     | N/A  |
|       └─Selection_33(Probe)          | 100.00  | 17408   | cop[tikv] |                          | time:101.3ms, loops:30, cop_task: {num: 14, max: 26.5ms, min: 2.33ms, avg: 10.7ms, p95: 26.5ms, max_proc_keys: 5103, p95_proc_keys: 5103, tot_proc: 127ms, rpc_num: 14, rpc_time: 149.9ms, copr_cache: disabled}, tikv_task:{proc max:23ms, min:0s, p80:17ms, p95:23ms, iters:80, tasks:14}, scan_detail: {total_process_keys: 20001, total_keys: 20001, rocksdb: {delete_skipped_count: 0, key_skipped_count: 0, block: {cache_hit_count: 5825, read_count: 0, read_byte: 0 Bytes}}} | gt(test.t.c, 2000)                 | N/A     | N/A  |
|         └─TableRowIDScan_32          | 122.32  | 20001   | cop[tikv] | table:t                  | tikv_task:{proc max:23ms, min:0s, p80:17ms, p95:23ms, iters:80, tasks:14}, scan_detail: {total_process_keys: 0, total_keys: 0, rocksdb: {delete_skipped_count: 0, key_skipped_count: 0, block: {cache_hit_count: 0, read_count: 0, read_byte: 0 Bytes}}}                                                                                                                                                                                                                              | keep order:false                   | N/A     | N/A  |
+--------------------------------------+---------+---------+-----------+--------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------+---------+------+
7 rows in set (0.059 sec)
```




![image](https://user-images.githubusercontent.com/14054293/112273378-3ec68200-8cb8-11eb-84e7-3c95355b481b.png)
QPS up ten times

### CODE REVIEW

File: executor/distsql.go [hunk 1]
```diff
@@ -379,13 +385,40 @@ func (e *IndexLookUpExecutor) Open(ctx context.Context) error {
...
+	if !e.openCloseByInside {
+		e.originRange = e.ranges
+	}
+	if e.indexSidePagination {
+		switch e.paginationSize {
+		case -1: // run too many times, remove Limit executor and read all rest data.
+			e.dagPB.Executors = e.dagPB.Executors[:len(e.dagPB.Executors)-1]
+			e.indexSidePagination = false
+		case 0: // add Limit executor, init page size.
+			e.paginationSize = 1
```
Comment (coocood) [inline executor/distsql.go:397]: 1 is too small.

Comment (wshwsh12) [reply executor/distsql.go:397]: ONLY FOR TEST

File: util/ranger/ranger.go [hunk 2]
```diff
@@ -433,6 +433,50 @@ func UnionRanges(sc *stmtctx.StatementContext, ranges []*Range, mergeConsecutive
 	return ranges, nil
 }
 
+// SplitRange split one range by encode key.
+// For ranges ([a, b], [c, d]) we have guaranteed that a <= c.
+// If a <= val <= b. this two range can be split as ([a, val]),([val, b], [c, d]).
+func SplitRange(sc *stmtctx.StatementContext, ranges []*Range, val []types.Datum) (left []*Range, right []*Range, err error) {
```
Comment (XuHuaiyu) [inline util/ranger/ranger.go:439]: GetNextRangeByLastKey

Comment (XuHuaiyu) [inline util/ranger/ranger.go:439]: val --> lastKey

Comment (XuHuaiyu) [inline util/ranger/ranger.go:439]: Maybe we can pass `desc` as an input argument of this function?

File: executor/distsql.go [hunk 3]
```diff
@@ -611,6 +653,32 @@ func (e *IndexLookUpExecutor) Next(ctx context.Context, req *chunk.Chunk) error
...
+		return err
+	}
+	// rebuild ranges
+	left, right, err := ranger.SplitRange(e.ctx.GetSessionVars().StmtCtx, e.ranges, e.lastRowKeys)
+	if e.desc {
+		e.ranges = left
+	} else {
+		e.ranges = right
+	}
+	if err != nil {
```
Comment (XuHuaiyu) [inline executor/distsql.go:673]: move this check to line 668?

File: util/ranger/ranger.go [hunk 4]
```diff
@@ -433,6 +433,54 @@ func UnionRanges(sc *stmtctx.StatementContext, ranges []*Range, mergeConsecutive
 	return ranges, nil
 }
 
+// GetNextRangeByLastKey get next Range by last key.
+// For ranges ([a, b], [c, d]) we have guaranteed that a <= c.
+// If a <= val <= b. this two range can be split as ([a, val]),([val, b], [c, d]).
+func GetNextRangeByLastKey(sc *stmtctx.StatementContext, ranges []*Range, lastKey []types.Datum, desc bool) ([]*Range, error) {
```
Comment (winoros) [inline util/ranger/ranger.go:439]: Why `LastKey` is not just one?

Comment (wshwsh12) [reply util/ranger/ranger.go:439]: []types.Datum is one key..

File: executor/builder.go [hunk 5]
```diff
@@ -2934,15 +2942,15 @@ func buildNoRangeIndexLookUpReader(b *executorBuilder, v *plannercore.PhysicalIn
 		// Should output pid col.
 		handleLen++
 	}
-	indexReq, indexStreaming, err := buildIndexReq(b, len(is.Index.Columns), handleLen, v.IndexPlans)
+	needIndexPaging := v.TryIndexPaging && is.KeepOrder
```
Comment (winoros) [inline executor/builder.go:2945]: The `KeepOrder` should be true when the `TryIndexPaging` is true?

File: executor/distsql.go [hunk 6]
```diff
@@ -600,6 +639,9 @@ func (e *IndexLookUpExecutor) Next(ctx context.Context, req *chunk.Chunk) error
 			return err
 		}
 		if resultTask == nil {
+			if e.needIndexPaging {
+				goto paginate
```
Comment (winoros) [inline executor/distsql.go:643]: Could we just wrap the goto part to a single method?

File: planner/core/task.go [hunk 7]
```diff
@@ -900,6 +900,9 @@ func (p *PhysicalLimit) sinkIntoIndexLookUp(t task) bool {
 	// We can sink Limit into IndexLookUpReader only if tablePlan contains no Selection.
 	ts, isTableScan := reader.tablePlan.(*PhysicalTableScan)
 	if !isTableScan {
+		if p.ctx.GetSessionVars().IndexLookupSize%10 == 1 {
```
Comment (winoros) [inline planner/core/task.go:903]: Why we check `%10` here?

File: executor/distsql.go [hunk 8]
```diff
@@ -379,13 +386,34 @@ func (e *IndexLookUpExecutor) Open(ctx context.Context) error {
...
+	if !e.isImplicitClose {
+		e.originRange = e.ranges
+		e.pageSize = int64(e.ctx.GetSessionVars().IndexLookupSize)
+	}
+	if e.needIndexPaging {
+		if !e.hasAddLimit {
+			e.dagPB.Executors = append(e.dagPB.Executors, e.constructLimitPB(uint64(e.pageSize)))
+			e.hasAddLimit = true
+		} else {
+			e.pageSize = e.pageSize * 2
```
Comment (winoros) [inline executor/distsql.go:398]: When the size reaches the `BatchSize`, the limit is not used anymore. Can we just fall back to the default logic?

File: planner/core/physical_plans.go [hunk 9]
```diff
@@ -260,6 +260,8 @@ type PhysicalIndexLookUpReader struct {
 	ExtraHandleCol *expression.Column
 	// PushedLimit is used to avoid unnecessary table scan tasks of IndexLookUpReader.
 	PushedLimit *PushedDownLimit
+	// TryIndexPaging try to page IndexScan, since there is a Limit can't push down to IndexScan.
+	TryIndexPaging bool
```
Comment (winoros) [inline planner/core/physical_plans.go:264]: Can we move it to the execution phase only?

File: planner/core/task.go (hunk 7 above)
Comment (wshwsh12) [reply planner/core/task.go:903]: It is easy to open/close the hot-fix.

File: executor/distsql.go (hunk 8 above)
Comment (wshwsh12) [reply executor/distsql.go:398]: We need stop IndexScan in time.

File: executor/distsql.go [hunk 10]
```diff
@@ -678,6 +750,21 @@ type indexWorker struct {
 	PushedLimit *plannercore.PushedDownLimit
 }
 
+func (w *indexWorker) constructLookUpContent(row chunk.Row, handleIdx []int) []types.Datum {
```
Comment (coocood) [inline executor/distsql.go:753]: It's actually index key, not lookup key. Should name it`constructIndexKey`

File: executor/distsql.go [hunk 11]
```diff
@@ -345,6 +345,13 @@ type IndexLookUpExecutor struct {
 	PushedLimit *plannercore.PushedDownLimit
 
 	stats *IndexLookUpRunTimeStats
+
+	needIndexPaging bool
+	hasAddLimit     bool
+	pageSize        int64
+	isImplicitClose bool
+	originRange     []*ranger.Range
+	lastRowKeys     []types.Datum
```
Comment (coocood) [inline executor/distsql.go:354]: lastIndexKey is better.

**Pull request closed without merging**